### ISSUE TITLE
Bug: [no-duplicate-type-constituents] Shouldn't report on error types

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#9576]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=C4TwDgpgBAKlC8UCCUA%2ByDcAoIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false

### Repro Code

```TypeScript
type T = A | A;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-duplicate-type-constituents": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

When types are _error_ types, they shouldn't be factored into rules that check type equality.

### Actual Result

We still get an error even though the types are errors.

### Additional Info

Previously mentioned by @bradzacher in https://github.com/typescript-eslint/typescript-eslint/issues/9344#issuecomment-2164770946.

💖
